airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -0,0 +1,153 @@
+import functools
+from abc import ABC, abstractmethod
+from typing import Any, List, Mapping, Optional, Protocol, Tuple
+
+from airbyte_cdk.sources.connector_state_manager import ConnectorStateManager
+from airbyte_cdk.sources.message import MessageRepository
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+
+
+def _extract_value(mapping: Mapping[str, Any], path: List[str]) -> Any:
+    return functools.reduce(lambda a, b: a[b], path, mapping)
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "Comparable", other: "Comparable") -> bool:
+        pass
+
+
+class CursorField:
+    def __init__(self, path: List[str]) -> None:
+        self._path = path
+
+    def extract_value(self, record: Record) -> Comparable:
+        return _extract_value(record.data, self._path)  # type: ignore  # we assume that the value the path points at is a comparable
+
+
+class Cursor(ABC):
+    @abstractmethod
+    def observe(self, record: Record) -> None:
+        """
+        Indicate to the cursor that the record has been emitted
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def close_partition(self, partition: Partition) -> None:
+        """
+        Indicate to the cursor that the partition has been successfully processed
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def end_sync(self) -> None:
+        """
+        Indicate to the cursor that the sync has been successfully processed
+        """
+        raise NotImplementedError()
+
+
+class NoopCursor(Cursor):
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: Partition) -> None:
+        pass
+
+    def end_sync(self) -> None:
```

Comment:
I had to add this method. The reason is that a connector might have multiple partitions that are not generated because of the incremental part. The "per partition" example:
* Given the slices `{"parent_id": "1"}` and `{"parent_id": "2"}`. The stream could be incremental but the `most_recend_record` is not accurate in terms of defining the new state (we might need to update `observe` to consider the partition at some point)

So now, the new behavior is:
* when we have boundaries within the slice, we update the state at every partition
* when we don't have boundaries and the partition has an identifier, we break as we don't support "per partition" yet
* else, we wait for the sync to end successfully before updating based on the most recent record

Reply:
This is not needed as we can update the state using the most recent record in `close_partition` as well

Reply:
Good point, we should document that. Here is one case that shows that it is a possibility : https://github.com/airbytehq/airbyte/blob/1ce84d6396e446e1ac2377362446e3fb94509461/airbyte-integrations/connectors/source-stripe/source_stripe/streams.py#L363-L379